### ISSUE TITLE
runs executed from the CLI don't show up in Runs tab of job page

### ISSUE THREAD

**Issue by sryza** [dagster-io/dagster#17707]
### To reproduce

Use this module:
```python
from dagster import Definitions, asset, define_asset_job


@asset
def asset17704() -> None:
    ...


defs = Definitions(assets=[asset17704], jobs=[define_asset_job("job17704")])
```

Execute from the CLI:
`dagster job execute -f <module_path> -j job17704`

Observe that the asset is marked as materialized on the asset graph and the global Runs page shows the run.

Click into the job page for job17704 and click into the Runs tab. Observe that it's empty.

**Comment by hellendag** [comment]
Is it possible that there is a tag missing from the run created by the CLI?

**Comment by sryza** [comment]
I just tried comparing a run created from the CLI and with a run created from the UI.  Neither of them appear to have any tags.

Aside from tags, are there any differences in how the frontend filters/queries for runs on the Job page vs. runs on the Runs page?

**Comment by salazarm** [comment]
@sryza the only difference is that we include the job/pipeline name in the query https://github.com/dagster-io/dagster/blob/master/js_modules/dagster-ui/packages/ui-core/src/pipelines/PipelineRunsRoot.tsx#L99

**Comment by salazarm** [comment]
Okay I figured out its due to the UI adding a [filter for the repository ](https://github.com/dagster-io/dagster/pull/10786/files#diff-4f62b46fb6f68e9b085144a40216ff760bb4f9df2bfb576ab4302a618a96616dR70-R76)

I believe it's necessary so likely we need to fix this by making sure that tag gets set by the CLI entrypoint

**Comment by sryza** [comment]
Taking the UI tag off of this because it seems like the solution involves the CLI entrypoint.

We probably want to recreate this logic https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/dagster_run.py#L382-L384 inside https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_cli/job.py.

**Comment by maximearmstrong** [comment]
I've been investigating this - it seems to be working just fine for `dagster job launch`, but not for `dagster job execute`.

For `dagster job launch`, we create the `DagsterRun` in [dagster.dagster._cli.job](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_cli/job.py), to which we pass the `external_job_origin` parameter, which is needed to add the repository tag in [dagster.dagster._core.storage.dagster_run](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/dagster_run.py#L382-L384).

The problem with `dagster job execute` is that the `DagsterRun` is not created in [dagster.dagster._cli.job](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_cli/job.py), but later in [dagster.dagster._core.execution.api](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/execution/api.py). Namely, the repository tag logic already exists for `dagster job execute`, in `DagsterRun`, but we don't pass the `external_job_origin` parameter to `execute_job`, see [here](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/execution/api.py#L334-L343).

This [is the code](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_cli/job.py#L459-L467) to extract the `external_job_origin` - we need the `WorkspaceRequestContext`. 

Potential solutions:
1- We find a way to pass the `WorkspaceRequestContext` to `execute_job`, through the `ReconstructableJob` maybe? 
2- We have the `DagsterInstance` in `execute_job`, so we could also pass the the Dagster version and kwargs to execute job and recreate [this logic](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_cli/job.py#L459-L467) there.

WDYT @sryza?

**Comment by gibsondan** [comment]
cc @srya just as a sanity check there - loading the whole workspace just to calculate some tags is a bit odd, but i think worth it if it will address this problem.

(The main difference between dagster job execute and dagster job launch is that the former runs the job in-process and the latter launches it using the run launcher defined on the instance - which is why its a bit odd to load the workspace here, since usually that is only loaded so that you can then use it to generate snapshots without running the code directly - but here, we are about to run the code directly anyways)

**Comment by sryza** [comment]
When you say "loading the whole workspace", do you mean triggering import of the Python modules for each code location? Or just parsing the workspace.yaml to see the list of code locations? Seems like the former could be pretty expensive? Is it required?

Btw this is not necessarily a super high priority issue. So if it requires a bunch of heavy lifting or has undesirable side effects, could make sense to table it for now.

**Comment by gibsondan** [comment]
The former - the main reason it would be required would be to figure out what repository to use if you didn't specify a repository name in the CLI, which requires loading the code to see what repositories are there.

**Comment by sryza** [comment]
If only we could be confident that everyone was using `Definitions`

**Comment by sryza** [comment]
Seems like we'd only need to load one code location though, which is the code location that includes the job, not all code location in the workspace, right?

**Comment by gibsondan** [comment]
That's true (unless you want to get fancy and try to match the job name to the only code location with a job of that name) - it probably should load the location if there is only one and raise an exception if its ambiguous (i.e. there is more than one location and you didn't specify a location arg to scope it down)
